build fails with napoleon enabled: Null object is not iterable

### Problem
Experimenting with converting [scons](https://github.com/SCons/scons) to generate API docs using Sphinx instead of Epydoc. During this, someone asked if could turn on display of class members starting with underscores. Experimenting with this by enabling napoleon extension and adding the two flag settings as described in http://www.sphinx-doc.org/en/master/ext/example_google.html

napoleon_include_special_with_doc = True
napoleon_include_private_with_doc = True

and the build blew up. I backed off the two flags, and the build still blew up.  Since the error message requested filing an issue on this, I am doing so.  I don't have a "simple reproducer", my simple example using the extension works fine.  Removing the line for the napoleon extension allows the build to complete.

#### Error Results
```
$ make html
Running Sphinx v1.5.6
loading pickled environment... not yet created
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 13 source files that are out of date
updating environment: 13 added, 0 changed, 0 removed
reading sources... [ 38%] SCons.Script                                          
Exception occurred:
  File "/usr/lib/python3.6/site-packages/sphinx/ext/napoleon/__init__.py", line 421, in _skip_member
    cls_path, _, _ = qualname.rpartition('.')
TypeError: 'Null' object is not iterable
The full traceback has been saved in /tmp/sphinx-err-qxul9qji.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
make: *** [Makefile:20: html] Error 1
```

#### Error log
[sphinx-err-qxul9qji.log](https://github.com/sphinx-doc/sphinx/files/1967804/sphinx-err-qxul9qji.log)

### Environment info
- OS: Fedora 26
- Sphinx version: 1.5.6
- Python version: 3.6.5 (CPython)
- Docutils version: 0.13.1 release
- Jinja2 version: 2.9.6

--- a/CHANGES
+++ b/CHANGES
@@ -40,6 +40,7 @@ Bugs fixed
 * #4980: latex: Explicit labels on code blocks are duplicated
 * #4919: node.asdom() crashes if toctree has :numbered: option
 * #4914: autodoc: Parsing error when using dataclasses without default values
+* #4931: autodoc: crashed when handler for autodoc-skip-member raises an error
 
 Testing
 --------

--- a/CHANGES
+++ b/CHANGES
@@ -41,6 +41,8 @@ Bugs fixed
 * #4919: node.asdom() crashes if toctree has :numbered: option
 * #4914: autodoc: Parsing error when using dataclasses without default values
 * #4931: autodoc: crashed when handler for autodoc-skip-member raises an error
+* #4931: autodoc: crashed when subclass of mocked class are processed by
+  napoleon module
 
 Testing
 --------

--- a/sphinx/ext/autodoc/__init__.py
+++ b/sphinx/ext/autodoc/__init__.py
@@ -25,7 +25,7 @@ from sphinx.deprecation import RemovedInSphinx20Warning
 from sphinx.ext.autodoc.importer import mock, import_object, get_object_members
 from sphinx.ext.autodoc.importer import _MockImporter  # to keep compatibility  # NOQA
 from sphinx.ext.autodoc.inspector import format_annotation, formatargspec  # to keep compatibility  # NOQA
-from sphinx.locale import _
+from sphinx.locale import _, __
 from sphinx.pycode import ModuleAnalyzer, PycodeError
 from sphinx.util import logging
 from sphinx.util import rpartition, force_decode
@@ -643,11 +643,17 @@ class Documenter(object):
             # should be skipped
             if self.env.app:
                 # let extensions preprocess docstrings
-                skip_user = self.env.app.emit_firstresult(
-                    'autodoc-skip-member', self.objtype, membername, member,
-                    not keep, self.options)
-                if skip_user is not None:
-                    keep = not skip_user
+                try:
+                    skip_user = self.env.app.emit_firstresult(
+                        'autodoc-skip-member', self.objtype, membername, member,
+                        not keep, self.options)
+                    if skip_user is not None:
+                        keep = not skip_user
+                except Exception as exc:
+                    logger.warning(__('autodoc: failed to determine %r to be documented.'
+                                      'the following exception was raised:\n%s'),
+                                   member, exc)
+                    keep = False
 
             if keep:
                 ret.append((membername, member, isattr))

--- a/sphinx/ext/autodoc/importer.py
+++ b/sphinx/ext/autodoc/importer.py
@@ -23,7 +23,7 @@ from sphinx.util.inspect import isenumclass, safe_getattr
 
 if False:
     # For type annotation
-    from typing import Any, Callable, Dict, Generator, List, Optional, Tuple  # NOQA
+    from typing import Any, Callable, Dict, Generator, Iterator, List, Optional, Tuple  # NOQA
 
 logger = logging.getLogger(__name__)
 
@@ -41,7 +41,7 @@ class _MockObject(object):
 
     def __init__(self, *args, **kwargs):
         # type: (Any, Any) -> None
-        pass
+        self.__qualname__ = ''
 
     def __len__(self):
         # type: () -> int
@@ -52,8 +52,8 @@ class _MockObject(object):
         return False
 
     def __iter__(self):
-        # type: () -> None
-        pass
+        # type: () -> Iterator
+        return iter([])
 
     def __mro_entries__(self, bases):
         # type: (Tuple) -> Tuple

